File: android-app/app/src/main/java/org/opentrafficmap/citslogger/MiniMqttClient.java
package org.opentrafficmap.citslogger;

import javax.net.SocketFactory;
import javax.net.ssl.SSLSocketFactory;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.Closeable;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Locale;

final class MiniMqttClient implements Closeable {
    private static final int KEEPALIVE_SECONDS = 60;

    private Socket socket;
    private InputStream in;
    private BufferedOutputStream out;
    private Thread pingThread;
    private volatile boolean running;
    private String topicPrefix;

    synchronized void connect(String mqttUri, String nodeId, String hardwareVariant, String firmwareVersion) throws Exception {
        close();
        if (nodeId == null || nodeId.trim().isEmpty()) throw new IllegalArgumentException("Node ID is empty");

        URI uri = URI.create(mqttUri.trim());
        String scheme = uri.getScheme() == null ? "mqtt" : uri.getScheme().toLowerCase(Locale.ROOT);
        boolean tls;
        int defaultPort;
        if ("mqtt".equals(scheme)) {
            tls = false;
            defaultPort = 1883;
        } else if ("mqtts".equals(scheme) || "ssl".equals(scheme)) {
            tls = true;
            defaultPort = 8883;
        } else {
            throw new IllegalArgumentException("Use mqtt:// or mqtts:// URI");
        }
        String host = uri.getHost();
        if (host == null || host.isEmpty()) throw new IllegalArgumentException("MQTT host is empty");
        int port = uri.getPort() > 0 ? uri.getPort() : defaultPort;

        SocketFactory factory = tls ? SSLSocketFactory.getDefault() : SocketFactory.getDefault();
        socket = factory.createSocket(host, port);
        socket.setTcpNoDelay(true);
        socket.setKeepAlive(true);
        socket.setSoTimeout(15000);
        in = socket.getInputStream();
        out = new BufferedOutputStream(socket.getOutputStream(), 64 * 1024);

        topicPrefix = "its/" + nodeId.trim() + "/";
        String clientId = "cits-android-" + nodeId.trim();
        String user = null;
        String pass = null;
        if (uri.getUserInfo() != null) {
            int colon = uri.getUserInfo().indexOf(':');
            if (colon >= 0) {
                user = uri.getUserInfo().substring(0, colon);
                pass = uri.getUserInfo().substring(colon + 1);
            } else {
                user = uri.getUserInfo();
            }
        }

        sendConnect(clientId, topicPrefix + "status", "offline".getBytes(StandardCharsets.UTF_8), user, pass);
        readConnAck();
        running = true;
        startPingThread();

        publish(topicPrefix + "status", "online".getBytes(StandardCharsets.UTF_8), true, false);
        publish(topicPrefix + "info", makeInfoPayload(nodeId, hardwareVariant, firmwareVersion), false, false);
        publish(topicPrefix + "stats", "{}".getBytes(StandardCharsets.UTF_8), false, true);
    }

    synchronized void publishPacket(byte[] payload) throws IOException {
        if (!isConnected()) throw new IOException("MQTT is not connected");
        publish(topicPrefix + "packet", payload, false, false);
    }

    synchronized void flush() throws IOException {
        if (out != null) out.flush();
    }

    synchronized boolean isConnected() {
        return running && socket != null && socket.isConnected() && !socket.isClosed();
    }

    private byte[] makeInfoPayload(String nodeId, String hardwareVariant, String firmwareVersion) {
        String emac = nodeId;
        if (nodeId.matches("(?i)[0-9a-f]{12}")) {
            emac = nodeId.substring(0, 2) + ":" + nodeId.substring(2, 4) + ":" +
                    nodeId.substring(4, 6) + ":" + nodeId.substring(6, 8) + ":" +
                    nodeId.substring(8, 10) + ":" + nodeId.substring(10, 12);
        }
        String hwv = safeJson(hardwareVariant == null || hardwareVariant.isEmpty() ? "android-bridge" : hardwareVariant);
        String ver = safeJson(firmwareVersion == null || firmwareVersion.isEmpty() ? "android-bridge" : firmwareVersion);
        String json = "{\"emac\":\"" + safeJson(emac) + "\",\"ver\":\"" + ver + "\",\"hwv\":\"" + hwv + "\"}";
        return json.getBytes(StandardCharsets.UTF_8);
    }

    private static String safeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private void sendConnect(String clientId, String willTopic, byte[] willPayload,
                             String username, String password) throws IOException {
        ByteArrayOutputStream vh = new ByteArrayOutputStream();
        writeUtf8(vh, "MQTT");
        vh.write(4);
        int flags = 0x02 | 0x04 | 0x20; // clean session, will flag, retained will; QoS 0
        if (password != null) flags |= 0x40;
        if (username != null) flags |= 0x80;
        vh.write(flags);
        writeShort(vh, KEEPALIVE_SECONDS);

        ByteArrayOutputStream payload = new ByteArrayOutputStream();
        writeUtf8(payload, clientId);
        writeUtf8(payload, willTopic);
        writeBinary(payload, willPayload);
        if (username != null) writeUtf8(payload, username);
        if (password != null) writeUtf8(payload, password);

        ByteArrayOutputStream body = new ByteArrayOutputStream();
        body.write(vh.toByteArray());
        body.write(payload.toByteArray());
        writePacket(0x10, body.toByteArray(), true);
    }

    private void readConnAck() throws IOException {
        DataInputStream din = new DataInputStream(in);
        int header = din.readUnsignedByte();
        if (header != 0x20) throw new IOException("Expected CONNACK, got 0x" + Integer.toHexString(header));
        int rem = readRemainingLength(din);
        if (rem != 2) throw new IOException("Bad CONNACK length " + rem);
        int flags = din.readUnsignedByte();
        int code = din.readUnsignedByte();
        if (code != 0) throw new IOException("MQTT CONNACK refused, code " + code + ", flags " + flags);
    }

    private synchronized void publish(String topic, byte[] payload, boolean retain, boolean flush) throws IOException {
        if (out == null) throw new IOException("MQTT is not connected");
        ByteArrayOutputStream body = new ByteArrayOutputStream();
        writeUtf8(body, topic);
        body.write(payload);
        writePacket(retain ? 0x31 : 0x30, body.toByteArray(), flush);
    }

    private void writePacket(int fixedHeader, byte[] body, boolean flush) throws IOException {
        out.write(fixedHeader);
        writeRemainingLength(out, body.length);
        out.write(body);
        if (flush) out.flush();
    }

    private void startPingThread() {
        pingThread = new Thread(() -> {
            while (running) {
                try {
                    Thread.sleep(KEEPALIVE_SECONDS * 500L);
                    synchronized (MiniMqttClient.this) {
                        if (out != null && running) {
                            out.write(0xC0);
                            out.write(0x00);
                            out.flush();
                        }
                    }
                } catch (Exception ignored) {
                    running = false;
                    break;
                }
            }
        }, "mqtt-ping");
        pingThread.setDaemon(true);
        pingThread.start();
    }

    private static void writeUtf8(ByteArrayOutputStream out, String value) throws IOException {
        byte[] b = value.getBytes(StandardCharsets.UTF_8);
        writeShort(out, b.length);
        out.write(b);
    }

    private static void writeBinary(ByteArrayOutputStream out, byte[] value) throws IOException {
        writeShort(out, value.length);
        out.write(value);
    }

    private static void writeShort(ByteArrayOutputStream out, int value) {
        out.write((value >>> 8) & 0xff);
        out.write(value & 0xff);
    }

    private static void writeRemainingLength(OutputStream out, int value) throws IOException {
        do {
            int encodedByte = value % 128;
            value /= 128;
            if (value > 0) encodedByte |= 128;
            out.write(encodedByte);
        } while (value > 0);
    }

    private static int readRemainingLength(DataInputStream in) throws IOException {
        int multiplier = 1;
        int value = 0;
        int encodedByte;
        do {
            encodedByte = in.readUnsignedByte();
            value += (encodedByte & 127) * multiplier;
            multiplier *= 128;
            if (multiplier > 128 * 128 * 128) throw new IOException("Malformed MQTT remaining length");
        } while ((encodedByte & 128) != 0);
        return value;
    }

    @Override
    public synchronized void close() {
        running = false;
        try {
            if (out != null) {
                out.write(0xE0);
                out.write(0x00);
                out.flush();
            }
        } catch (Exception ignored) {}
        try { if (socket != null) socket.close(); } catch (Exception ignored) {}
        socket = null;
        in = null;
        out = null;
    }
}

File: android-app/app/src/main/java/org/opentrafficmap/citstogo/bridge/MiniMqttClient.kt
package org.opentrafficmap.citstogo.bridge

import java.io.BufferedOutputStream
import java.io.ByteArrayOutputStream
import java.io.Closeable
import java.io.DataInputStream
import java.io.IOException
import java.io.OutputStream
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Locale
import javax.net.SocketFactory
import javax.net.ssl.SSLSocketFactory

class MiniMqttClient : Closeable {
    private var socket: java.net.Socket? = null
    private var input: DataInputStream? = null
    private var output: BufferedOutputStream? = null
    private var pingThread: Thread? = null
    @Volatile private var running = false
    private var topicPrefix = ""

    @Synchronized
    fun connect(mqttUri: String, nodeId: String, appVersion: String) {
        close()
        val cleanNode = nodeId.trim()
        require(cleanNode.isNotEmpty()) { "Node ID is empty" }
        val uri = parseUri(mqttUri)
        val tls = when (uri.scheme.lowercase(Locale.ROOT)) {
            "mqtt" -> false
            "mqtts", "ssl" -> true
            else -> throw IllegalArgumentException("Use mqtt:// or mqtts:// URI")
        }
        val host = uri.host ?: throw IllegalArgumentException("MQTT host is empty")
        val port = if (uri.port > 0) uri.port else if (tls) 8883 else 1883
        val factory: SocketFactory = if (tls) SSLSocketFactory.getDefault() else SocketFactory.getDefault()
        val s = factory.createSocket(host, port)
        s.tcpNoDelay = true
        s.keepAlive = true
        s.soTimeout = 15_000
        socket = s
        input = DataInputStream(s.getInputStream())
        output = BufferedOutputStream(s.getOutputStream(), 64 * 1024)

        topicPrefix = "its/$cleanNode/"
        val userInfo = uri.rawUserInfo?.let { java.net.URLDecoder.decode(it, StandardCharsets.UTF_8.name()) }
        val username = userInfo?.substringBefore(':', missingDelimiterValue = userInfo)
        val password = userInfo?.substringAfter(':', missingDelimiterValue = "")
            ?.takeIf { userInfo.contains(':') }
        val clientId = "cits-android-${clientSafe(cleanNode)}"

        sendConnect(clientId, topicPrefix + "status", "offline".toByteArray(), username, password)
        readConnAck()
        running = true
        startPingThread()
        publish(topicPrefix + "status", "online".toByteArray(), retain = true, flush = true)
        publish(topicPrefix + "info", infoPayload(cleanNode, appVersion), retain = false, flush = true)
        publishStats(0)
    }

    @Synchronized
    fun publishPacket(payload: ByteArray) {
        checkConnected()
        publish(topicPrefix + "packet", payload, retain = false, flush = false)
    }

    @Synchronized
    fun publishStats(secondsSinceStart: Long) {
        checkConnected()
        val json = """{"rbt":$secondsSinceStart}"""
        publish(topicPrefix + "stats", json.toByteArray(), retain = false, flush = true)
    }

    @Synchronized
    fun flush() {
        output?.flush()
    }

    fun isConnected(): Boolean {
        val s = socket
        return running && s != null && s.isConnected && !s.isClosed
    }

    @Synchronized
    override fun close() {
        running = false
        runCatching {
            output?.write(0xE0)
            output?.write(0)
            output?.flush()
        }
        runCatching { socket?.close() }
        socket = null
        input = null
        output = null
    }

    private fun sendConnect(
        clientId: String,
        willTopic: String,
        willPayload: ByteArray,
        username: String?,
        password: String?,
    ) {
        val variableHeader = ByteArrayOutputStream()
        writeUtf8(variableHeader, "MQTT")
        variableHeader.write(4)
        var flags = 0x02 or 0x04 or 0x20
        if (password != null) flags = flags or 0x40
        if (username != null) flags = flags or 0x80
        variableHeader.write(flags)
        writeShort(variableHeader, KEEPALIVE_SECONDS)

        val payload = ByteArrayOutputStream()
        writeUtf8(payload, clientId)
        writeUtf8(payload, willTopic)
        writeBinary(payload, willPayload)
        if (username != null) writeUtf8(payload, username)
        if (password != null) writeUtf8(payload, password)

        val body = ByteArrayOutputStream()
        body.write(variableHeader.toByteArray())
        body.write(payload.toByteArray())
        writePacket(0x10, body.toByteArray(), flush = true)
    }

    private fun readConnAck() {
        val input = input ?: throw IOException("MQTT input is closed")
        val header = input.readUnsignedByte()
        if (header != 0x20) throw IOException("Expected CONNACK, got 0x${header.toString(16)}")
        val remaining = readRemainingLength(input)
        if (remaining != 2) throw IOException("Bad CONNACK length $remaining")
        val flags = input.readUnsignedByte()
        val code = input.readUnsignedByte()
        if (code != 0) throw IOException("MQTT CONNACK refused, code $code, flags $flags")
    }

    private fun publish(topic: String, payload: ByteArray, retain: Boolean, flush: Boolean) {
        val body = ByteArrayOutputStream()
        writeUtf8(body, topic)
        body.write(payload)
        writePacket(if (retain) 0x31 else 0x30, body.toByteArray(), flush)
    }

    private fun writePacket(fixedHeader: Int, body: ByteArray, flush: Boolean) {
        val output = output ?: throw IOException("MQTT output is closed")
        output.write(fixedHeader)
        writeRemainingLength(output, body.size)
        output.write(body)
        if (flush) output.flush()
    }

    private fun startPingThread() {
        pingThread = Thread({
            while (running) {
                try {
                    Thread.sleep(KEEPALIVE_SECONDS * 500L)
                    synchronized(this) {
                        val out = output
                        if (running && out != null) {
                            out.write(0xC0)
                            out.write(0)
                            out.flush()
                        }
                    }
                } catch (_: Exception) {
                    running = false
                    break
                }
            }
        }, "mqtt-ping").apply {
            isDaemon = true
            start()
        }
    }

    private fun checkConnected() {
        if (!isConnected()) throw IOException("MQTT is not connected")
    }

    private fun infoPayload(nodeId: String, appVersion: String): ByteArray {
        val emac = if (nodeId.matches(Regex("(?i)[0-9a-f]{12}"))) {
            nodeId.chunked(2).joinToString(":")
        } else {
            nodeId
        }
        val json = """{"emac":"${jsonEscape(emac)}","ver":"${jsonEscape(appVersion)}","hwv":"android-usb-bridge"}"""
        return json.toByteArray(StandardCharsets.UTF_8)
    }

    companion object {
        private const val KEEPALIVE_SECONDS = 60

        private fun parseUri(raw: String): URI {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) throw IllegalArgumentException("MQTT URI is empty")
            return URI.create(if ("://" in trimmed) trimmed else "mqtt://$trimmed")
        }

        private fun clientSafe(value: String): String =
            URLEncoder.encode(value, StandardCharsets.UTF_8.name()).take(48)

        private fun writeUtf8(out: ByteArrayOutputStream, value: String) {
            val bytes = value.toByteArray(StandardCharsets.UTF_8)
            writeShort(out, bytes.size)
            out.write(bytes)
        }

        private fun writeBinary(out: ByteArrayOutputStream, value: ByteArray) {
            writeShort(out, value.size)
            out.write(value)
        }

        private fun writeShort(out: ByteArrayOutputStream, value: Int) {
            out.write((value ushr 8) and 0xff)
            out.write(value and 0xff)
        }

        private fun writeRemainingLength(out: OutputStream, value: Int) {
            var remaining = value
            do {
                var encodedByte = remaining % 128
                remaining /= 128
                if (remaining > 0) encodedByte = encodedByte or 128
                out.write(encodedByte)
            } while (remaining > 0)
        }

        private fun readRemainingLength(input: DataInputStream): Int {
            var multiplier = 1
            var value = 0
            var encodedByte: Int
            do {
                encodedByte = input.readUnsignedByte()
                value += (encodedByte and 127) * multiplier
                multiplier *= 128
                if (multiplier > 128 * 128 * 128) throw IOException("Malformed MQTT remaining length")
            } while ((encodedByte and 128) != 0)
            return value
        }

        private fun jsonEscape(value: String): String =
            value.replace("\\", "\\\\").replace("\"", "\\\"")
    }
}
